From 84c6c08a6d5380e72ddf36cc2aff184e4d4ae295 Mon Sep 17 00:00:00 2001
From: Dan Kephart <dan.kephart@lairdtech.com>
Date: Fri, 4 Dec 2015 17:43:03 -0500
Subject: [PATCH] ath6kl: Always use vmalloc for allocating memory to store
 firmware

Always allocate memory using vmalloc to store the firmware loaded from
the filesystem.  This avoids a vfree error and memory leak on module
unloading for firmware loaded using api1 and testmode finctions.

Bug: 7544
---
 drivers/net/wireless/ath/ath6kl/init.c | 27 +++++++++++++++++++++++++--
 1 file changed, 25 insertions(+), 2 deletions(-)

diff --git a/drivers/net/wireless/ath/ath6kl/init.c b/drivers/net/wireless/ath/ath6kl/init.c
index 284ab3c..9c06e03 100644
--- a/drivers/net/wireless/ath/ath6kl/init.c
+++ b/drivers/net/wireless/ath/ath6kl/init.c
@@ -690,6 +690,29 @@ static int ath6kl_get_fw(struct ath6kl *ar, const char *filename,
 	return ret;
 }
 
+static int ath6kl_get_fw_vm(struct ath6kl *ar, const char *filename,
+			 u8 **fw, size_t *fw_len)
+{
+	const struct firmware *fw_entry;
+	int ret;
+
+	ret = request_firmware(&fw_entry, filename, ar->dev);
+	if (ret)
+		return ret;
+
+	*fw_len = fw_entry->size;
+	*fw = vmalloc(*fw_len);
+
+	if (*fw == NULL)
+		ret = -ENOMEM;
+
+	memcpy(*fw, fw_entry->data, *fw_len);
+
+	release_firmware(fw_entry);
+
+	return ret;
+}
+
 #ifdef CONFIG_OF
 /*
  * Check the device tree for a board-id and use it to construct
@@ -836,7 +859,7 @@ static int ath6kl_fetch_testmode_file(struct ath6kl *ar)
 
 	set_bit(TESTMODE, &ar->flag);
 
-	ret = ath6kl_get_fw(ar, filename, &ar->fw, &ar->fw_len);
+	ret = ath6kl_get_fw_vm(ar, filename, &ar->fw, &ar->fw_len);
 	if (ret) {
 		ath6kl_err("Failed to get testmode %d firmware file %s: %d\n",
 			   ar->testmode, filename, ret);
@@ -861,7 +884,7 @@ static int ath6kl_fetch_fw_file(struct ath6kl *ar)
 	snprintf(filename, sizeof(filename), "%s/%s",
 		 ar->hw.fw.dir, ar->hw.fw.fw);
 
-	ret = ath6kl_get_fw(ar, filename, &ar->fw, &ar->fw_len);
+	ret = ath6kl_get_fw_vm(ar, filename, &ar->fw, &ar->fw_len);
 	if (ret) {
 		ath6kl_err("Failed to get firmware file %s: %d\n",
 			   filename, ret);
-- 
1.9.1

